native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -118,30 +119,68 @@ public static Object convertToJSON(BXml xml, String attributePrefix, boolean pre
+        Object children = convertBXmlSequence((BXmlSequence) xmlItem.getChildrenSeq(), attributePrefix,
+                preserveNamespaces, attributeManager);
+        if (mapData.size() > 0) {
+            if (children == null || children instanceof BString) {
+                if (children != null) {
+                    putAsBStrings(mapData, CONTENT, children.toString().trim());
+                    rootNode = mapData;
+                } else {
+                    putAsBStrings(rootNode, keyValue, mapData);
+                }
             } else {
-                BMap<BString, Object> attrMap = newJsonMap();
-                addAttributes(attrMap, attributePrefix, attributeMap);
-
-                putAsBStrings(rootNode, keyValue, attrMap);
+                BMap<BString, Object> data = (BMap<BString, Object>) children;
+                for (Map.Entry<BString, Object> entry: mapData.entrySet()) {
+                    data.put(entry.getKey(), entry.getValue());
+                }
+                putAsBStrings(rootNode, keyValue, data);
+            }
+        } else {
+            if (children == null) {
+                putAsBStrings(rootNode, keyValue, EMPTY_STRING);
+            } else if (children instanceof BString) {
+                putAsBStrings(rootNode, keyValue, children.toString().trim());
+            } else {
+                putAsBStrings(rootNode, keyValue, children);
             }
-            return rootNode;
         }
-
-        putAsBStrings(rootNode, keyValue, children);
-        addAttributes(rootNode, attributePrefix, attributeMap);
         return rootNode;
     }
 
+    private static void processAttributes(BXmlItem xmlItem, boolean preserveNamespaces, String attributePrefix,
+                                      BMap<BString, Object> mapData, AttributeManager attributeManager) {
+        LinkedHashMap<String, String> tempAttributeMap =  new LinkedHashMap<>();
+        if (attributeManager.getMap().isEmpty()) {
+            tempAttributeMap = collectAttributesAndNamespaces(xmlItem, preserveNamespaces);
+            attributeManager.initializeMap(tempAttributeMap);
+        } else {
+            LinkedHashMap<String, String> newAttributeMap = collectAttributesAndNamespaces(xmlItem,
+                    preserveNamespaces);
+            if (!newAttributeMap.isEmpty()) {
+                for (Map.Entry<String, String> entrySet : newAttributeMap.entrySet()) {
+                    String key = entrySet.getKey();
+                    String value = entrySet.getValue();
+                    LinkedHashMap<String, String> attributeMap = attributeManager.getMap();
+                    if (attributeMap.get(key) == null || !attributeMap.get(key).equals(value)) {
+                        attributeManager.setValue(key, value);
+                        tempAttributeMap.put(key, value);
+                    }
+                }
+            }
+        }
+        if (!tempAttributeMap.isEmpty()) {
```

Comment:
Same, we don't need to call is empty or not if there is a for each loop at the next step